xray/src/generation.rs

```diff
@@ -127,7 +128,27 @@ pub trait ColoringStrategy: Send {
     // the pixel (x, y) in the final tile image.
     fn get_pixel_color(&self, x: u32, y: u32, background_color: Color<u8>) -> Color<u8>;
 
-    fn attributes(&self) -> Vec<&'static str>;
+    fn attributes(&self) -> HashSet<String>;
+}
+
+trait BinnedColoringStrategy {
+    fn binning(&self) -> &Binning;
+    fn bins(&self, points_batch: &PointsBatch) -> Vec<i64> {
+        if let Some((attrib_name, size)) = self.binning() {
+            if let Some(attr_data) = points_batch.attributes.get(attrib_name) {
+                macro_rules! rhs {
+                    ($dtype:ident, $data:ident, $size:expr) => {
+                        $data
+                            .iter()
+                            .map(|e| *e as i64 / *$size as i64)
```

Comment:
Wouldn't we want something like `(*e as f64 / *$size) as i64` here? Right now I think this will panic if `size < 1` (I don't know if you want to suport such a use case).

Reply:
Good point, I changed it.